src/main/scala/mesosphere/marathon/state/AppDefinition.scala

```diff
@@ -365,6 +366,38 @@ case class AppDefinition(
     val baseId = id.canonicalPath(base)
     copy(id = baseId, dependencies = dependencies.map(_.canonicalPath(baseId)))
   }
+
+  def portAssignments(task: Task): Option[Seq[PortAssignment]] = {
+    def fromIpAddress: Option[Seq[PortAssignment]] = ipAddress.flatMap {
+      case IpAddress(_, _, DiscoveryInfo(appPorts)) =>
+        for {
+          launched <- task.launched
+          effectiveIpAddress <- task.effectiveIpAddress(this)
+        } yield appPorts.zip(launched.hostPorts).zipWithIndex.map {
+          case ((appPort, hostPort), portIndex) =>
+            PortAssignment(Some(appPort.name), portIndex, effectiveIpAddress, hostPort)
+        }.toList
+    }
+
+    def fromPortMappings: Option[Seq[PortAssignment]] =
+      for {
+        pms <- portMappings
+        launched <- task.launched
+      } yield pms.zip(launched.hostPorts).zipWithIndex.map {
+        case ((portMapping, hostPort), portIndex) =>
+          PortAssignment(portMapping.name, portIndex, task.agentInfo.host, hostPort)
+      }.toList
+
+    def fromPortDefinitions: Option[Seq[PortAssignment]] = task.launched.map { launched =>
+      portDefinitions.zip(launched.hostPorts).zipWithIndex.map {
+        case ((portDefinition, hostPort), portIndex) =>
+          PortAssignment(portDefinition.name, portIndex, task.agentInfo.host, hostPort)
+      }
+    }
+
+    if (ipAddress.isDefined) fromIpAddress // If an IP-per-Task was requested, don't fall back
+    else fromPortMappings.orElse(fromPortDefinitions)
```

Comment:
Answering myself: it doesn't make any sense, but Marathon allows you to create an app with `bridged` networking, no port mappings and multiple host ports, like this:

```json
{
  "id": "foo",
  "cmd": null,
  "cpus": 1,
  "mem": 128,
  "disk": 0,
  "instances": 1,
  "container": {
    "docker": {
      "image": "python:2.7-alpine",
      "network": "BRIDGE"
    },
    "type": "DOCKER"
  },
  "ports": [1,2,3]
}
```